lib/fuse.c

```diff
@@ -2967,6 +2967,20 @@ static void fuse_lib_unlink(fuse_req_t req, fuse_ino_t parent,
 		fuse_prepare_interrupt(f, req, &d);
 		if (!f->conf.hard_remove && is_open(f, parent, name)) {
 			err = hide_node(f, path, parent, name);
+			if (!err) {
+				/* we have hidden the node so now check again under a lock in case it is not used any more */
+				if (!is_open(f, parent, wnode->name)) {
+					char *unlinkpath;
+
+					/* get the hidden file path, to unlink it */
+					if (try_get_path(f, wnode->nodeid, NULL, &unlinkpath, NULL, false) == 0) {
+						err = fuse_fs_unlink(f->fs, unlinkpath);
+						if (!err)
+							remove_node(f, parent, wnode->name);
+						free(unlinkpath);
+					}
```

Comment:
Make sense,  but I'm worried about `remove_node`, what is if someone opened the 'hidden' file in the mean time?  I _think_ one can open these .fuse_hidden files?

Reply:
Unless I am mistaken, the path should be locked by the previous `get_path_wrlock()`.
An `fuse_lib_open()` of the hidden file will wait for `fuse_lib_unlink()` to finish.

Reply:
Ah right, current unlink also calls remove_node - if there would be a race, it would have been there already.